accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -1236,18 +1246,92 @@ the `LEGACY_ERASURE` of the types.
 
 ### Const evaluation and canonicalization
 
-With unsound null checking, all generic const constructors and generic
-const literals are treated as if all type arguments passed to them were legacy
-types (both at the top level, and recursively over the structure of the
-types), regardless of whether the constructed class is defined in a legacy
-library or not, and regardless of whether the constructor invocation or
-literal occurs in a legacy library or not.  This ensures that const objects
-continue to canonicalize consistently across legacy and opted-in libraries.
+Const evaluation is modified so that legacy and opted-in instances canonicalize
+more consistently as defined below, using the operation **LEGACY_TYPE**(`T`) on
+types `T` defined as follows.
+
+- **LEGACY_TYPE**(`T`) = `T` if `T` is `dynamic`, `void`, `Null`
+- **LEGACY_TYPE**(`T`) = `T*` if `T` is `Never` or `Object`
+- **LEGACY_TYPE**(`FutureOr<T>`) = `FutureOr<S>*`
+- **LEGACY_TYPE**(`T?`) =
+  - let `S` be **LEGACY_TYPE**(`T`)
+  - if `S` is `R*` then `R?`
+  - else `S?`
+- **LEGACY_TYPE**(`T*`) = **LEGACY_TYPE**(`T`)
+- **LEGACY_TYPE**(`X extends T`) = `X*`
+- **LEGACY_TYPE**(`X & T`) =
+  - This case should not occur, since intersection types are not permitted as
+    generic arguments.
+- **LEGACY_TYPE**(`C<T0, ..., Tn>`) = `C<R0, ..., Rn>*`
+  - where `Ri` is **LEGACY_TYPE**(`Ti`)
+- **LEGACY_TYPE**(`R Function<X extends B>(S)`) = `F*`
+  - where `F = R1 Function<X extends B1>(S1)`
+  - and `R1` = **LEGACY_TYPE**(`R`)
+  - and `B1` = **LEGACY_TYPE**(`B`)
+  - and `S1` = **LEGACY_TYPE**(`S`)
```

Comment:
Shouldn't `required` markers be removed from the signature? They are removed by LEGACY_ERASURE, which is informally specified in the [spec](https://github.com/dart-lang/language/blob/master/accepted/future-releases/nnbd/feature-specification.md#classes-defined-in-legacy-libraries). For clarity, shouldn't LEGACY_ERASURE be specified in the same way as the new LEGACY_TYPE?

Reply:
@crelier I don't think it should.  Informally, my goal is that two constant type literals which are "spelled the same" should compare identical.  More formally, my goal is that two constant type literals should compare identical iff they are equal according to the runtime type equality definition specified just above this.  That definition does ignore `*` modifiers, but does not ignore `required` markers.   That was the idea anyway, do you see issues with it?  It does bother me a bit that I have two definitions  (LEGACY_TYPE and LEGACY_ERASURE) that are so close but not the same.  

And yes, I should have specified LEGACY_ERASURE precisely too, if I get a chance I'll follow up.

Reply:
@leafpetersen Thanks for the clarification! I was just checking, because the VM currently applies LEGACY_ERASURE on constants when  reading them in, and therefore removes the markers. My understanding is that the VM will now apply LEGACY_TYPE instead, so the markers will now stay.